drgn/helpers/linux/idr.py

```diff
@@ -7,21 +7,89 @@
 
 The ``drgn.helpers.linux.idr`` module provides helpers for working with the IDR
 data structure in :linux:`include/linux/idr.h`. An IDR provides a mapping from
-an ID to a pointer. This currently only supports Linux v4.11+; before this,
-IDRs were not based on radix trees.
+an ID to a pointer.
 """
 
+import operator
 from typing import Iterator, Tuple
 
-from _drgn import _linux_helper_idr_find as idr_find
-from drgn import Object
+from _drgn import _linux_helper_idr_find as idr_find_in_radix_tree
+from drgn import NULL, IntegerLike, Object, Program, cast, sizeof
 from drgn.helpers.linux.radixtree import radix_tree_for_each
 
 __all__ = (
     "idr_find",
     "idr_for_each",
 )
 
+_IDR_BITS = 8
+_IDR_MASK = (1 << _IDR_BITS) - 1
+
+
+def _max_idr_shift(prog: Program) -> int:
+    return sizeof(prog.type("int")) * 8 - 1
+
+
+def _max_idr_level(prog: Program) -> int:
+    return (_max_idr_shift(prog) + _IDR_BITS - 1) // _IDR_BITS
+
+
+def _round_up(number: int, multiple: int) -> int:
+    return -((-number) // multiple) * multiple
+
+
+def _idr_max(prog: Program, layers: int) -> int:
+    bits = min(layers * _IDR_BITS, _max_idr_shift(prog))
+    return (1 << bits) - 1
+
+
+def _fls(num: int) -> int:
+    num_bits = 32
+    if num == 0:
+        return 0
+
+    for i in range(num_bits):
+        if (num & 1) == 1:
+            pos = i
+
+        num = num >> 1
+
+    return pos + 1
+
+
+def idr_find(idr: Object, id: IntegerLike) -> Object:
+    """
+    Lookup entry with the given ID in an IDR.
+
+    :param idr: ``struct idr *``
+    :param id: ``Entry ID``
+    :return: ``void *`` found entry, or NULL if not found.
+    """
+    id = operator.index(id)
+
+    if hasattr(idr, "idr_rt"):
```

Comment:
We should add a comment mentioning the exact commit that made the change here and some additional explanation.

```suggestion
    # Since Linux kernel commit 0a835c4f090a ("Reimplement IDR and IDA using
    # the radix tree") (in v4.11), IDRs are backed by radix trees. Before that,
    # they are a separate data structure. The helper in libdrgn only handles
    # the radix tree version.
    if hasattr(idr, "idr_rt"):
```